phy/phy.nim

```diff
@@ -0,0 +1,346 @@
+    # it's a single declaration
+    syntaxCheck(code, source_checks, decl)
+    ctx.declToIL(code, NodeIndex(0))
+  of ExprNodes:
+    # it's a standalone expression
+    syntaxCheck(code, source_checks, expr)
+    discard ctx.exprToIL(code)
+  of spec_source.NodeKind.Module:
+    # it's a full module. Translate all declarations
+    syntaxCheck(code, source_checks, module)
+    for it in code.items(NodeIndex(0)):
+      ctx.declToIL(code, it)
+
+    # the last procedure is the one that will be executed
+  else:
+    error "unexpected node: " & $code[NodeIndex(0)].kind
+
+  # don't continue if there was an error:
+  let messages = reporter[].retrieve()
+  if messages.len > 0:
+    for it in messages.items:
+      echo "Error: ", it
+    quit(2)
+
+  result[1] =
+    if ctx.procList.len > 0: ctx.procList[^1].result
+    else:                    prim(tkError)
+  result[0] = ctx.close()
+
+proc compile(tree: var PackedTree[spec.NodeKind], source, target: Language) =
+  assert source != langSource
+  assert ord(source) >= ord(target)
+  var current = source
+  while current != target:
+    case current
+    of lang0, langBytecode, langSource:
+      assert false, "cannot be handled here: " & $current
+    of lang1:
+      syntaxCheck(tree, lang1_checks, module)
+      # TODO: don't hardcode the pointer size
+      tree = tree.apply(pass1.lower(tree, 8))
+      current = lang0
+    of lang3:
+      syntaxCheck(tree, lang3_checks, module)
+      # TODO: don't hardcode the pointer size
+      tree = tree.apply(pass3.lower(tree, 8))
+      current = lang1
+    of lang4:
+      syntaxCheck(tree, lang4_checks, module)
+      tree = tree.apply(pass4.lower(tree))
+      current = lang3
+    of lang10:
+      syntaxCheck(tree, lang10_checks, module)
+      tree = tree.apply(pass10.lower(tree))
+      current = lang4
+
+    print(tree, current)
+
+proc main(args: openArray[string]) =
+  var opts = initOptParser(args)
+  var input = ""
+
+  var source = langSource
+  var target = langBytecode
+
+  var cmd = None
+
+  for (kind, key, arg) in opts.getopt():
+    case kind
+    of cmdShortOption:
+      error "unknown option: " & key
+    of cmdLongOption:
+      case key
+      of "runner":
+        gRunner = true
+      of "source":
+        source = parseEnum[Language](arg)
+      of "target":
+        target = parseEnum[Language](arg)
+      of "show":
+        gShow.incl parseEnum[Language](arg)
+      else:
+        error "unknown option: " & key
+    of cmdArgument:
+      if cmd == None:
+        cmd = parseEnum[Command](key)
+      else:
+        # must be the input filename
+        input = key
+    of cmdEnd:
+      unreachable()
+
+  # make sure the command line arguments are sensible:
+  case cmd
+  of None:
+    echo HelpText
+    quit(1)
+  of Eval:
+    if target != langBytecode:
+      error "'e' requires the target language to be '" & $langBytecode & "'"
+  else:
+    discard "nothing special to do"
+
+  if ord(target) > ord(source):
```

Comment:
If I'm reading this correctly, everything above this line is CLI interfacing, what follows is basically all "driver" logic that runs the requested file/test. So if it makes sense to allow reusing the logic in this module then this line and below need to be lifted into its own routine.